File: ruby/examples/number-clamp_test.rb
# カード number-clamp の Contract を検証するテスト（Comparable#clamp）
require "minitest/autorun"
require "date"

class NumberClampTest < Minitest::Test
  def test_clamps_to_bounds_inclusive
    # 範囲外は境界値に、範囲内はそのまま。境界値は含む
    assert_equal 100, 120.clamp(0, 100)
    assert_equal 0, -5.clamp(0, 100)
    assert_equal 42, 42.clamp(0, 100)
    assert_equal 0, 0.clamp(0, 100)
    assert_equal 100, 100.clamp(0, 100)
  end

  def test_range_form_and_endless_ranges
    # Range 版と、端無し Range による片側のみの制限
    assert_equal 100, 120.clamp(0..100)
    assert_equal 42, 42.clamp(0..100)
    assert_equal 100, 120.clamp(..100)
    assert_equal(-5, -5.clamp(..100))
    assert_equal 0, -5.clamp(0..)
    assert_equal 120, 120.clamp(0..)
  end

  def test_returns_argument_object_without_conversion
    # 返り値は self / min / max のオブジェクトそのもので、型変換しない
    x = 5
    assert_same x, x.clamp(0, 10)
    assert_equal 9.5, 10.clamp(0.5, 9.5)
    assert_instance_of Float, 10.clamp(0.5, 9.5)
    assert_instance_of Integer, 5.clamp(0.5, 9.5)
  end

  def test_min_greater_than_max_raises
    # min > max は ArgumentError
    assert_raises(ArgumentError) { 5.clamp(10, 1) }
    assert_raises(ArgumentError) { 5.clamp(10..1) }
  end

  def test_exclusive_range_with_end_raises_but_endless_is_fine
    # 終端がある排他的な Range は ArgumentError。終端の無い 0... は 0.. と同じ
    assert_raises(ArgumentError) { 5.clamp(0...10) }
    assert_raises(ArgumentError) { 5.clamp(...10) }
    assert_equal 5, 5.clamp(0...)
    assert_equal 0, -5.clamp(0...)
  end

  def test_nil_bound_means_unbounded
    # min / max に nil を渡すとその側の制限を設けない
    assert_equal 5, 5.clamp(nil, 10)
    assert_equal 120, 120.clamp(0, nil)
    assert_equal 5, 5.clamp(nil, nil)
  end

  def test_nan_raises_instead_of_returning_nan
    # NaN が混ざると ArgumentError
    assert_raises(ArgumentError) { Float::NAN.clamp(0, 100) }
    assert_raises(ArgumentError) { 5.clamp(Float::NAN, 100) }
    assert_raises(ArgumentError) { 200.clamp(0, Float::NAN) }
  end

  def test_incomparable_types_raise
    # 比較できない型は ArgumentError
    assert_raises(ArgumentError) { 5.clamp(0, "a") }
  end

  def test_single_non_range_argument_raises_type_error
    # Pitfalls: 引数 1 個は Range のみ。数値 1 個は TypeError
    assert_raises(TypeError) { 5.clamp(10) }
  end

  def test_works_for_any_comparable
    # Alternatives: 文字列や Date にもそのまま使える
    assert_equal "c", "z".clamp("a", "c")
    assert_equal Date.new(2024, 2, 1), Date.new(2024, 1, 1).clamp(Date.new(2024, 2, 1), Date.new(2024, 3, 1))
  end
end

File: python/examples/number-clamp_test.py
# カード number-clamp の Contract を検証するテスト
import math

import pytest


def clamp(x, lo, hi):
    return min(max(x, lo), hi)


def test_clamps_into_range_inclusive():
    """範囲外は境界値に置き換え、境界値は含む"""
    assert clamp(120, 0, 100) == 100
    assert clamp(-5, 0, 100) == 0
    assert clamp(42, 0, 100) == 42
    assert clamp(0, 0, 100) == 0
    assert clamp(100, 0, 100) == 100


def test_lo_greater_than_hi_always_returns_hi():
    """lo > hi なら常に hi が返る"""
    assert clamp(50, 100, 0) == 0
    assert clamp(-10, 100, 0) == 0
    assert clamp(200, 100, 0) == 0


def test_returns_one_of_the_arguments_without_conversion():
    """返り値は引数のいずれかそのもので、型変換しない"""
    x = 5
    assert clamp(x, 0, 10) is x
    assert type(clamp(5, 0.0, 10)) is int
    assert type(clamp(-5, 0.0, 10)) is float


def test_nan_handling():
    """x が NaN なら NaN。lo / hi の片方が NaN ならその境界だけが効かず、もう片方は効く"""
    assert math.isnan(clamp(math.nan, 0, 100))
    assert clamp(50, math.nan, 100) == 50
    assert clamp(-5, math.nan, 100) == -5
    assert clamp(200, math.nan, 100) == 100  # hi は効く
    assert clamp(200, 0, math.nan) == 200
    assert clamp(-5, 0, math.nan) == 0  # lo は効く


def test_does_not_mutate_arguments():
    """引数を変更しない（イミュータブルな数値なので値が変わらないことを確認）"""
    x, lo, hi = 120, 0, 100
    clamp(x, lo, hi)
    assert (x, lo, hi) == (120, 0, 100)


def test_incomparable_types_raise_type_error():
    """比較できない型を渡すと TypeError"""
    with pytest.raises(TypeError):
        clamp("a", 0, 1)


def test_alternatives():
    """Alternatives: max(lo, min(x, hi)) は lo > hi で lo を返す。sorted は lo <= hi なら同じ"""
    assert max(0, min(120, 100)) == 100
    assert max(100, min(50, 0)) == 100
    assert sorted((120, 0, 100))[1] == 100
    assert sorted((50, 100, 0))[1] == 50


def test_no_builtin_clamp():
    """Pitfalls: math モジュールに clamp は無い"""
    assert not hasattr(math, "clamp")
